### DISCUSSION TITLE
Implement order/then_order_by for combination clauses (Union, …)

### DISCUSSION THREAD

**Discussion by Aethelflaed** [diesel-rs/diesel#4246] [Ideas]
Postgres, mysql and sqlite all support ordering a combination query.

We could allow this by implementing `OrderDsl` / `ThenOrderDsl` for `CombinationClause`.

From what I understand, it would need to verify that the order's `expr` is valid for a given query source (as it's only implemented for `SelectStatement<FromClause<F>, ...>`).

It seems possible to mimic this constraint by implementing the traits only if the combinator's `Source` or `Rhs` is a query source (in which case the combinator itself becomes a query source)

I'm keen on working on that after https://github.com/diesel-rs/diesel/discussions/3084 if the idea is sound

**Comment by weiznich** [comment]
I fear that's currently not easily possible as the "subqueries" could have different select expressions with different field types. Making that possible would require a considerable amount of design work and likely some other features like `alias!` support for single fields first.

What should be relatively easy to implement is the fix for `positional_order_by` as that only requires changing what structure is generated.

**Reply by Aethelflaed** [reply]
I'll start with positional_order_by then

Edit: https://github.com/diesel-rs/diesel/pull/4251
